### PR TITLE
try use table.clone copy table

### PR DESCRIPTION
**Pull request by juzipeek**
we can use  luajit's table.clone func create a new nodes.

### CODE REVIEW

File: lib/resty/chash.lua [hunk 1]
```diff
@@ -38,12 +38,32 @@ void chash_point_delete(chash_point_t *old_points, uint32_t old_length,
...
+local new_tab
+do
+    local ok
+    ok, new_tab = pcall(require, "table.new")
+    if not ok or type(new_tab) ~= "function" then
+        new_tab = function (narr, nrec) return {} end
+    end
+end
+
+local function _copy(nodes)
```
Comment (spacewander) [inline lib/resty/chash.lua]: The `_copy` should go inside the `local copy` block.

File: lib/resty/chash.lua [hunk 2]
```diff
@@ -38,12 +38,32 @@ void chash_point_delete(chash_point_t *old_points, uint32_t old_length,
...
+    local newnodes = new_tab(0, 5)
```
Comment (spacewander) [inline lib/resty/chash.lua]: Better to explain the reason you choice 5.

File: lib/resty/chash.lua [hunk 3]
```diff
@@ -99,6 +119,7 @@ local function _precompute(nodes)
     for id, weight in pairs(nodes) do
         newnodes[id] = weight
     end
```
Comment (spacewander) [inline lib/resty/chash.lua]: Maybe we don't need the lines above any more?

File: lib/resty/roundrobin.lua [hunk 4]
```diff
@@ -9,16 +9,41 @@ local math_random = math.random
 local _M = {}
 local mt = { __index = _M }
 
+local new_tab
```
Comment (spacewander) [inline lib/resty/roundrobin.lua]: Better to avoid repeating the code. We can put this stuff into a common module.

**Review by spacewander — CHANGES_REQUESTED** [review]

**Comment by spacewander** [comment]
@juzipeek 
Please take a look at the failed test cases.

File: lib/resty/chash.lua (hunk 1 above)
Comment (juzipeek) [reply lib/resty/chash.lua]: already move this code to utils module

File: lib/resty/chash.lua (hunk 3 above)
Comment (juzipeek) [reply lib/resty/chash.lua]: already delete this code

File: lib/resty/roundrobin.lua (hunk 4 above)
Comment (juzipeek) [reply lib/resty/roundrobin.lua]: already move this code to utils module

**Comment by juzipeek** [comment]
> @juzipeek
> Please take a look at the failed test cases.

1. new pull request change the test cases.
2. add OSX's dlib build support.

File: lib/resty/balancer/utils.lua [hunk 5]
```diff
@@ -0,0 +1,53 @@
+--
+-- first try use luajit's function
```
Comment (spacewander) [inline lib/resty/balancer/utils.lua:2]: The comment is confusing...
Is it suitable to appear at the top of this file?

File: lib/resty/balancer/utils.lua [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+--
+
+local _M = {}
+
+_M.name = "balancer-utils"
+_M.version = "0.03"
```
Comment (spacewander) [inline lib/resty/balancer/utils.lua:8]: The `name` and `version` are unused.

File: lib/resty/chash.lua [hunk 7]
```diff
@@ -58,6 +55,12 @@ local function load_shared_lib(so_name)
 
     local cpath = package.cpath
 
+    local postfix = ".so"
+    if ffi.os == "OSX" then
+        postfix = ".dylib"
+    end
+    so_name = so_name .. postfix
```
Comment (spacewander) [inline lib/resty/chash.lua:62]: I remember that when the ext name is missing, LuaJIT will automatically add it according to the platform. Therefore, we don't need to add it by ourselves.

File: t/roundrobin.t [hunk 8]
```diff
@@ -134,28 +133,16 @@ server2: 33334
...
-                ["server3"] = 1,
-                ["server4"] = 1,
-                ["server5"] = 1,
-                ["server6"] = 1,
-            }
 
-            rr:reinit(new_servers)
-            id = rr:find()
-            ngx.say(id)
+            local rr2 = roundrobin:new(servers, true)
```
Comment (spacewander) [inline t/roundrobin.t:137]: Better to use `reinit` like the original test.

File: t/roundrobin.t [hunk 9]
```diff
@@ -44,29 +44,19 @@ __DATA__
 
             for i = 1, 14 do
                 local id = rr:find()
-
-                ngx.say("id: ", id)
```
Comment (spacewander) [inline t/roundrobin.t:48]: For this sanity test case, maybe we can use three servers with the same weight, and check if they are roundrobined?

**Review by spacewander — CHANGES_REQUESTED** [review]

**Comment by spacewander** [comment]
> add OSX's dlib build support.

Please don't hijack the PR with another feature. We need to separate each feature in their own pull request.

File: lib/resty/chash.lua (hunk 7 above)
Comment (rainingmaster) [reply lib/resty/chash.lua:62]: Maybe we can add macOS as a new test environment for this change? https://docs.travis-ci.com/user/reference/osx/

Comment (juzipeek) [reply lib/resty/chash.lua:62]: thank u, i remove the ext name.

File: t/roundrobin.t (hunk 9 above)
Comment (juzipeek) [reply t/roundrobin.t:48]: i plan use three servers with the same weight, check current find different with the last find.  is this ok?

File: lib/resty/balancer/utils.lua (hunk 6 above)
Comment (juzipeek) [reply lib/resty/balancer/utils.lua:8]: > > add OSX's dlib build support.
> 
> Please don't hijack the PR with another feature. We need to separate each feature in their own pull request.

i will split into two PR.

File: t/roundrobin.t (hunk 8 above)
Comment (juzipeek) [reply t/roundrobin.t:137]: ok

File: lib/resty/balancer/utils.lua (hunk 5 above)
Comment (juzipeek) [reply lib/resty/balancer/utils.lua:2]: i will remove this comment.

File: t/roundrobin.t (hunk 9 above)
Comment (spacewander) [reply t/roundrobin.t:48]: @juzipeek
This is OK.

**Comment by doujiang24** [comment]
@juzipeek thanks for your contribution, it looks good to me except these following comments:
1. as @spacewander said, we need to separate each feature in their own pull request.
2. we'd better not change the existing test case when the library behavior doesn't change.

**Comment by spacewander** [comment]
@doujiang24 
The behavior  is changed slightly, since the order of the hash table is different between `table.copy` and manual copying.

**Comment by doujiang24** [comment]
@juzipeek Thanks for your contribution, merged by splitting into two PRs.
https://github.com/openresty/lua-resty-balancer/pull/34
https://github.com/openresty/lua-resty-balancer/pull/33

**Pull request closed without merging**